Generate everything needed regarding a WebKit Secure Coding type (and move DDScannerResult over to the generated format)

beidson opened the pull request:
#### 9bd720b73a8f1fea62d0e743755d4a9bda8fffb8
<pre>
Generate everything needed regarding a WebKit Secure Coding type (and move DDScannerResult over to the generated format)
<a href="https://bugs.webkit.org/show_bug.cgi?id=265728">https://bugs.webkit.org/show_bug.cgi?id=265728</a>
<a href="https://rdar.apple.com/119079477">rdar://119079477</a>

Reviewed by Alex Christensen.

All WebKit Secure Coding types will follow the identical pattern of:
1 - Get property list dictionary
2 - Encode it
3 - Decode it
4 - Verify the decoded dictionary contents match the type
5 - Recreate the object with the property list

Instead of writing all of that code over time after time, let&apos;s generate it.

Given an Objective-C type and a description of its expected dictionary format, the generator now:
1 - Generates a CoreIPC wrapper for that type
2 - Generates the ArgumentCoders for that CoreIPC wrapper (including code that validates the dictionary format)
3 - Generates a description of the dictionary in SerializedTypeInfo
4 - Generates a description of the CoreIPC wrapper in SerializedTypeInfo

All of the above was enough to move DDScannerResult over to this new format, and will greatly
accelerate how quickly we can add new types.

* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Scripts/generate-serializers.py:
(SerializedType.__init__):
(SerializedType.namespace_and_name):
(SerializedType.name_declaration_for_serialized_type_info):
(SerializedType.members_for_serialized_type_info):
(SerializedType.cpp_type_from_struct_or_class):
(SerializedType):
(SerializedType.cpp_struct_or_class_name):
(SerializedType.is_webkit_secure_coding_type):
(SerializedType.wrapper_for_webkit_secure_coding_type):
(generate_forward_declarations):
(generate_impl):
(generate_one_serialized_type_info):
(output_sorted_headers):
(generate_serialized_type_info):
(parse_serialized_types):
(generate_one_dictionary_member_validation):
(generate_webkit_secure_coding_impl):
(generate_webkit_secure_coding_impl.is):
(generate_webkit_secure_coding_header):
(main):
(SerializedType.namespace_unless_wtf_and_name): Deleted.
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.cpp:
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.h:
* Source/WebKit/Scripts/webkit/tests/GeneratedWebKitSecureCoding.cpp: Added.
(WebKit::CoreIPCNSSomeFoundationType::CoreIPCNSSomeFoundationType):
(WebKit::CoreIPCNSSomeFoundationType::isValidDictionary):
(WebKit::CoreIPCNSSomeFoundationType::toID const):
(WebKit::CoreIPCDDScannerResult::CoreIPCDDScannerResult):
(WebKit::CoreIPCDDScannerResult::isValidDictionary):
(WebKit::CoreIPCDDScannerResult::toID const):
* Source/WebKit/Scripts/webkit/tests/GeneratedWebKitSecureCoding.h: Added.
(WebKit::CoreIPCNSSomeFoundationType::CoreIPCNSSomeFoundationType):
(WebKit::CoreIPCDDScannerResult::CoreIPCDDScannerResult):
* Source/WebKit/Scripts/webkit/tests/SerializedTypeInfo.cpp:
(WebKit::allSerializedTypes):
* Source/WebKit/Scripts/webkit/tests/TestSerializedType.serialization.in:
* Source/WebKit/Scripts/webkit/tests/WebKitPlatformGeneratedSerializers.cpp:
(IPC::ArgumentCoder&lt;WebKit::CoreIPCNSSomeFoundationType&gt;::encode):
(IPC::ArgumentCoder&lt;WebKit::CoreIPCNSSomeFoundationType&gt;::decode):
(IPC::ArgumentCoder&lt;WebKit::CoreIPCDDScannerResult&gt;::encode):
(IPC::ArgumentCoder&lt;WebKit::CoreIPCDDScannerResult&gt;::decode):
* Source/WebKit/Shared/Cocoa/CoreIPCDDScannerResult.h: Removed.
* Source/WebKit/Shared/Cocoa/CoreIPCDDScannerResult.mm: Removed.
* Source/WebKit/Shared/Cocoa/CoreIPCDDScannerResult.serialization.in:
* Source/WebKit/Shared/Cocoa/CoreIPCDictionary.h:
* Source/WebKit/Shared/Cocoa/CoreIPCDictionary.mm:
(WebKit::CoreIPCDictionary::CoreIPCDictionary):
(WebKit::CoreIPCDictionary::keyHasValueOfType const):
(WebKit::CoreIPCDictionary::keyIsMissingOrHasValueOfType const):
(WebKit::CoreIPCDictionary::collectionValuesAreOfType const):
(WebKit::CoreIPCDictionary::createNSDictionaryIfNeeded const):
(WebKit::CoreIPCDictionary::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCDictionary.serialization.in:
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.mm:
(WebKit::secureCodingValueFromID):
(WebKit::valueFromID):
* Source/WebKit/Shared/Cocoa/CoreIPCSecureCoding.h:
* Source/WebKit/Shared/Cocoa/CoreIPCSecureCoding.mm:
(WebKit::CoreIPCSecureCoding::conformsToWebKitSecureCoding):
(WebKit::CoreIPCSecureCoding::conformsToSecureCoding):
(WebKit::CoreIPCSecureCoding::CoreIPCSecureCoding):
* Source/WebKit/Shared/Cocoa/CoreIPCTypes.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: <a href="https://commits.webkit.org/271555@main">https://commits.webkit.org/271555@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/7c3dcf871a589129edcff5ab22bafa67fb8e5bca

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/28682 "12 style errors") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/7325 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/30053 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/31303 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/26174 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/9493 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/4686 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/26247 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/28952 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/6097 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/24659 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/5261 "Passed tests") | 
| [❌ 🧪 webkitpy](https://ews-build.webkit.org/#/builders/6/builds/28842 "webkitpy-tests (failure)") | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/5451 "Passed tests") | | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/32641 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/26295 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/26106 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/31659 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/5379 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/3550 "Passed tests") | [❌ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/29437 "Found 1 new API test failure: /WebKitGTK/TestWebKitUserContentFilterStore:/webkit/WebKitUserContentFilterStore/save-multiple-filters (failure)") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/6994 "Built successfully") | | | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6880 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/5846 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/5897 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCDDScannerResult.serialization.in:
```diff
@@ -20,12 +20,16 @@
+    P: Number
+    VN: Number
+    SR: ?Array<DDScannerResult>
+    V: ?String
```
I think the syntax would be more conventional if the question mark were at the end.
Also, are these 1-2 letter strings really the strings to use?  I kind of like variable names that indicate what they are for.

beidson replied:
This are literally the keys put in the dictionary by the other framework, so yes they are correct. We have no flexibility on that.

I can move the ? to the end though

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCDictionary.mm:
```diff
@@ -50,13 +51,76 @@
+bool CoreIPCDictionary::keyHasValueOfType(const String& key, IPC::NSType type) const
+{
+    createNSDictionaryIfNeeded();
+    id object = [m_nsDictionary.get() objectForKey:(NSString *)key];
```
(NSString *) is technically not necessary because of String::operator NSString *, but some people think it's nice.

beidson replied:
I'm some people!!!

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCDictionary.mm:
```diff
@@ -50,13 +51,76 @@
+    if (!object)
+        return true;
+
+    if (![object isKindOfClass:[NSDictionary class]])
```
I'm preferring dot syntax like NSDictionary.class more and more, but it doesn't really matter.

beidson replied:
I'm not against changing this. I use it sometimes, too (I think this was just a copy/paste from somewhere else.)

achristensen07 commented on Source/WebKit/Shared/Cocoa/CoreIPCDictionary.h:
```diff
@@ -58,6 +63,9 @@ class CoreIPCDictionary {
 
+    void createNSDictionaryIfNeeded() const;
+
+    mutable RetainPtr<NSDictionary> m_nsDictionary;
```
Instead of adding this, can we make ValueType a HashMap or just do linear searches on small finite sets of values?  Having an object that has a cached other form like this leads to error-prone code.

beidson replied:
The first code I wrote for this was making it a HashMap instead.
The problem that creates is having to write general hashing functions for all possible CoreIPCNSCFObject variants, which is actually why this has always been a Vector<KeyValuePairs>.

I agree on the point of the cached object leading to error prone code in general but the specifics here matter:
Since the validation step always happens on the decode side, and since we always need the NSDictionary on the decode side, leveraging the native Obj-C hashing seemed quite preferable.

achristensen07 commented on Source/WebKit/Scripts/webkit/tests/SerializedTypeInfo.cpp:
```diff
@@ -312,6 +313,78 @@ Vector<SerializedTypeInfo> allSerializedTypes()
+            },
+            {
+                "OptionalNumberKey"_s,
+                "?WebKit::CoreIPCNumber"_s
```
ditto, question mark at the end would be more conventional, I think.

beidson replied:
Sure.

achristensen07 commented on Source/WebKit/Scripts/webkit/tests/GeneratedWebKitSecureCoding.h:
```diff
@@ -0,0 +1,84 @@
+class CoreIPCDDScannerResult {
+public:
+    CoreIPCDDScannerResult(DDScannerResult *);
+    CoreIPCDDScannerResult(RetainPtr<DDScannerResult> object)
```
a const RetainPtr& would make less refcount churn.

beidson replied:
Yup!

achristensen07 commented on Source/WebKit/Scripts/generate-serializers.py:
```diff
@@ -106,6 +110,10 @@ def __init__(self, struct_or_class, cf_type, namespace, name, parent_class_name,
                         self.from_cf_method = value
                     if key == 'ForwardDeclaration':
                         self.forward_declaration = value
+                    if key == 'WebKitSecureCodingClass':
```
All the possible attributes currently have comment documentation at the top of this file.  Please add such documentation for all new attributes.

beidson replied:
Will do.

achristensen07 commented on Source/WebKit/Scripts/webkit/tests/GeneratedWebKitSecureCoding.cpp:
```diff
@@ -0,0 +1,119 @@
+
+RetainPtr<id> CoreIPCDDScannerResult::toID() const
+{
+    RELEASE_ASSERT([PAL::getDDScannerResultClass() instancesRespondToSelector:@selector(_initWithWebKitPropertyListData:)]);
```
I feel like we're going to need a call to PAL::isDataDetectorsCoreFrameworkAvailable() or we'll break the base system.

beidson replied:
I had thought of this and dug in a little.

tldr: Since the base system already never does any data detector operations, we'll already never be encoding or decoding these.

achristensen07 commented on Source/WebKit/Scripts/webkit/tests/GeneratedWebKitSecureCoding.h:
```diff
@@ -0,0 +1,84 @@
+    {
+    }
+
+    CoreIPCDictionary m_propertyList;
```
Maybe not for this type, but many types will need this to be a std::optional to preserve nullness.  That mistake has bitten me several times this year.

beidson replied:
I don't forsee that in this case:
If we're wrapping an object that actually exists, then by definition it cannot be null...
And then our contract for the webkit property list model is there will always be a dictionary.
The dictionary might be 100% optional keys and therefore empty, but the dictionary itself will be present.

If a situation comes up that breaks that assumption, we can change the generator then. But for now my approach is to only add the required features as they come up.

achristensen07 commented on Source/WebKit/Scripts/generate-serializers.py:
```diff
@@ -68,17 +71,20 @@ def specialization(self):
         self.from_cf_method = None
         self.forward_declaration = None
+        self.custom_secure_coding_class = None
+        self.custom_secure_coding_header = None
```
This doesn't need to be a member of the SerializedType class.  Headers are not necessarily associated with a type.  I think we should do something more like ConditionalHeader.webkit_platform for this instead.

achristensen07 commented on Source/WebKit/Scripts/generate-serializers.py:
```diff
@@ -1261,15 +1367,179 @@ def parse_serialized_types(file):
+    for type in serialized_types:
+        if not type.is_webkit_secure_coding_type():
+            continue
+        if type.custom_secure_coding_header is not None:
```
Then this isn't needed.  Just pass in a list of ConditionalHeaders for the secure coding .mm file.

beidson committed:
Generate everything needed regarding a WebKit Secure Coding type (and move DDScannerResult over to the generated format)
https://bugs.webkit.org/show_bug.cgi?id=265728
rdar://119079477

Reviewed by Alex Christensen.

All WebKit Secure Coding types will follow the identical pattern of:
1 - Get property list dictionary
2 - Encode it
3 - Decode it
4 - Verify the decoded dictionary contents match the type
5 - Recreate the object with the property list

Instead of writing all of that code over time after time, let's generate it.

Given an Objective-C type and a description of its expected dictionary format, the generator now:
1 - Generates a CoreIPC wrapper for that type
2 - Generates the ArgumentCoders for that CoreIPC wrapper (including code that validates the dictionary format)
3 - Generates a description of the dictionary in SerializedTypeInfo
4 - Generates a description of the CoreIPC wrapper in SerializedTypeInfo

All of the above was enough to move DDScannerResult over to this new format, and will greatly
accelerate how quickly we can add new types.

* Source/WebKit/DerivedSources-output.xcfilelist:
* Source/WebKit/DerivedSources.make:
* Source/WebKit/Scripts/generate-serializers.py:
(SerializedType.__init__):
(SerializedType.namespace_and_name):
(SerializedType.name_declaration_for_serialized_type_info):
(SerializedType.members_for_serialized_type_info):
(SerializedType.cpp_type_from_struct_or_class):
(SerializedType):
(SerializedType.cpp_struct_or_class_name):
(SerializedType.is_webkit_secure_coding_type):
(SerializedType.wrapper_for_webkit_secure_coding_type):
(generate_forward_declarations):
(generate_impl):
(generate_one_serialized_type_info):
(output_sorted_headers):
(generate_serialized_type_info):
(parse_serialized_types):
(generate_one_dictionary_member_validation):
(generate_webkit_secure_coding_impl):
(generate_webkit_secure_coding_impl.is):
(generate_webkit_secure_coding_header):
(main):
(SerializedType.namespace_unless_wtf_and_name): Deleted.
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.cpp:
* Source/WebKit/Scripts/webkit/tests/GeneratedSerializers.h:
* Source/WebKit/Scripts/webkit/tests/GeneratedWebKitSecureCoding.cpp: Added.
(WebKit::CoreIPCNSSomeFoundationType::CoreIPCNSSomeFoundationType):
(WebKit::CoreIPCNSSomeFoundationType::isValidDictionary):
(WebKit::CoreIPCNSSomeFoundationType::toID const):
(WebKit::CoreIPCDDScannerResult::CoreIPCDDScannerResult):
(WebKit::CoreIPCDDScannerResult::isValidDictionary):
(WebKit::CoreIPCDDScannerResult::toID const):
* Source/WebKit/Scripts/webkit/tests/GeneratedWebKitSecureCoding.h: Added.
(WebKit::CoreIPCNSSomeFoundationType::CoreIPCNSSomeFoundationType):
(WebKit::CoreIPCDDScannerResult::CoreIPCDDScannerResult):
* Source/WebKit/Scripts/webkit/tests/SerializedTypeInfo.cpp:
(WebKit::allSerializedTypes):
* Source/WebKit/Scripts/webkit/tests/TestSerializedType.serialization.in:
* Source/WebKit/Scripts/webkit/tests/WebKitPlatformGeneratedSerializers.cpp:
(IPC::ArgumentCoder<WebKit::CoreIPCNSSomeFoundationType>::encode):
(IPC::ArgumentCoder<WebKit::CoreIPCNSSomeFoundationType>::decode):
(IPC::ArgumentCoder<WebKit::CoreIPCDDScannerResult>::encode):
(IPC::ArgumentCoder<WebKit::CoreIPCDDScannerResult>::decode):
* Source/WebKit/Shared/Cocoa/CoreIPCDDScannerResult.h: Removed.
* Source/WebKit/Shared/Cocoa/CoreIPCDDScannerResult.mm: Removed.
* Source/WebKit/Shared/Cocoa/CoreIPCDDScannerResult.serialization.in:
* Source/WebKit/Shared/Cocoa/CoreIPCDictionary.h:
* Source/WebKit/Shared/Cocoa/CoreIPCDictionary.mm:
(WebKit::CoreIPCDictionary::CoreIPCDictionary):
(WebKit::CoreIPCDictionary::keyHasValueOfType const):
(WebKit::CoreIPCDictionary::keyIsMissingOrHasValueOfType const):
(WebKit::CoreIPCDictionary::collectionValuesAreOfType const):
(WebKit::CoreIPCDictionary::createNSDictionaryIfNeeded const):
(WebKit::CoreIPCDictionary::toID const):
* Source/WebKit/Shared/Cocoa/CoreIPCDictionary.serialization.in:
* Source/WebKit/Shared/Cocoa/CoreIPCNSCFObject.mm:
(WebKit::secureCodingValueFromID):
(WebKit::valueFromID):
* Source/WebKit/Shared/Cocoa/CoreIPCSecureCoding.h:
* Source/WebKit/Shared/Cocoa/CoreIPCSecureCoding.mm:
(WebKit::CoreIPCSecureCoding::conformsToWebKitSecureCoding):
(WebKit::CoreIPCSecureCoding::conformsToSecureCoding):
(WebKit::CoreIPCSecureCoding::CoreIPCSecureCoding):
* Source/WebKit/Shared/Cocoa/CoreIPCTypes.h:
* Source/WebKit/WebKit.xcodeproj/project.pbxproj:

Canonical link: https://commits.webkit.org/271555@main
